File: Unit 1/nCombination/nCombination.java
import java.util.ArrayList;
import java.util.List;

public class nCombination
{

    static List<List<Boolean>> comb(int n, List<Boolean> A, List<List<Boolean>> out)
    {
        if (n == 0)
        {
            out.add(new ArrayList<>(A)); // Creating a copy of the list
        } 
        else 
        {
            A.add(true);
            comb(n - 1, A, out);
            A.remove(A.size() - 1);

            A.add(false);
            comb(n - 1, A, out);
            A.remove(A.size() - 1);
        }

        return out;
    }

    public static void main(String[] args)
    {
        List<List<Boolean>> outx = new ArrayList<>();
        List<List<Boolean>> res = comb(3, new ArrayList<>(), outx);

        for (int i = 0; i < res.size(); i++)
        {
            System.out.print(i + " ");
            for (boolean value : res.get(i))
            {
                System.out.print(value + " ");
            }
            System.out.println();
        }
    }
}

File: Unit 1-Into/nCombination/nCombination.py
def comb(n:int, A:list, out:list):
    if n == 0:
        out.append(A)
    else:
        comb(n-1, A.append(True), out)
        comb(n-1, A.append(False), out)
    
    return out

outx = []
res = comb(3, [], outx)

for i in range(len(res)):
    print(i , res[i])
    # print(i , outx[i]) This line that in python when a list is passed in a function the function do not
    # create a copy of the function rather than that it start working on the list directy so whatever
    # changes are perform on the passed list directly the list in the main function
    
    
    